File: business/chatroom/src/main/java/io/agora/chatroom/activity/ChatroomListActivity.java
package io.agora.chatroom.activity;

import android.content.Intent;
import android.os.Bundle;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;
import com.alibaba.android.arouter.facade.annotation.Route;
import com.google.android.material.imageview.ShapeableImageView;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;
import io.agora.baseui.BaseActivity;
import io.agora.chatroom.R;
import io.agora.chatroom.fragment.ChatroomListFragment;
import io.agora.chatroom.model.PageViewModel;
import io.agora.config.RouterPath;
import io.agora.secnceui.widget.titlebar.ChatroomTitleBar;

@Route(path = RouterPath.ChatroomListPath)
public class ChatroomListActivity extends BaseActivity implements ChatroomTitleBar.OnBackPressListener, View.OnClickListener {

    private ChatroomTitleBar mVRTitleBar;
    private ShapeableImageView mVRAvatar;
    private ConstraintLayout mAvatarLayout;
    private ConstraintLayout mButtonLayout;
    private TabLayout mTableLayout;
    private ViewPager2 mViewPager;
    private PageViewModel pageViewModel;
    private TextView title;
    private int mCount;
    private int index;
    private int[] titles = {R.string.tab_layout_all,R.string.tab_layout_chat_room,R.string.tab_layout_audio_room};

    @Override
    protected int getLayoutId() {
        return R.layout.agora_room_list_layout;
    }

    @Override
    protected void initSystemFit() {
        setFitSystemForTheme(false, "#00000000");
        setStatusBarTextColor(false);
    }

    @Override
    protected void initView(Bundle savedInstanceState) {
        mVRTitleBar =  findViewById(R.id.title_bar);
        mVRAvatar = findViewById(R.id.image_avatar);
        mAvatarLayout = findViewById(R.id.avatar_layout);
        mTableLayout = findViewById(R.id.agora_tab_layout);
        mViewPager = findViewById(R.id.vp_fragment);
        mButtonLayout = findViewById(R.id.bottom_layout);
    }

    @Override
    protected void initData() {
        setupWithViewPager();
    }

    @Override
    protected void initListener() {
        pageViewModel = new ViewModelProvider(this).get(PageViewModel.class);

        mVRTitleBar.setOnBackPressListener(this);
        mButtonLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(ChatroomListActivity.this, ChatroomCreateActivity.class));
            }
        });
        mTableLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                if(tab.getCustomView() != null) {
                    index = tab.getPosition();
                    title = tab.getCustomView().findViewById(R.id.tab_item_title);
                    ViewGroup.LayoutParams layoutParams = title.getLayoutParams();
                    layoutParams.height = (int)dip2px(mContext, 26);
                    title.setTextSize(TypedValue.COMPLEX_UNIT_SP, 16);
                    title.setGravity(Gravity.CENTER);
                    String content = getString(titles[index]) + getString(R.string.room_tab_layout_count,String.valueOf(mCount));
                    title.setText(content);
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
                if(tab.getCustomView() != null) {
                    TextView title = tab.getCustomView().findViewById(R.id.tab_item_title);
                    title.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12);
                    title.setText(titles[tab.getPosition()]);
                }
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });

        mViewPager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                if (position == 0){
                    pageViewModel.setPageSelect(-1);
                }else if (position == 1){
                    pageViewModel.setPageSelect(0);
                }else if (position == 2){
                    pageViewModel.setPageSelect(1);
                }
            }
        });

        mVRAvatar.setOnClickListener(this);

    }

    private void setupWithViewPager() {
        mViewPager.setOffscreenPageLimit(ViewPager2.OFFSCREEN_PAGE_LIMIT_DEFAULT);
        // set adapter
        mViewPager.setAdapter(new FragmentStateAdapter(getSupportFragmentManager(), getLifecycle()) {
            @NonNull
            @Override
            public Fragment createFragment(int position) {
                ChatroomListFragment fragment = new ChatroomListFragment();
                Bundle bundle = new Bundle();
                bundle.putInt("position",position);
                fragment.setArguments(bundle);
                fragment.SetItemCountChangeListener(new ChatroomListFragment.itemCountListener() {
                    @Override
                    public void getItemCount(int count) {
                        mCount = count;
                        String content = getString(titles[index]) + getString(R.string.room_tab_layout_count,String.valueOf(count));
                        title.setText(content);
                    }
                });
                return fragment;
            }

            @Override
            public int getItemCount() {
                return titles.length;
            }

        });

        // set TabLayoutMediator
        TabLayoutMediator mediator = new TabLayoutMediator(mTableLayout, mViewPager, new TabLayoutMediator.TabConfigurationStrategy() {
            @Override
            public void onConfigureTab(@NonNull TabLayout.Tab tab, int position) {
                tab.setCustomView(R.layout.tab_item_layout);
                TextView title = tab.getCustomView().findViewById(R.id.tab_item_title);
                title.setText(titles[position]);
            }
        });
        // setup with viewpager2
        mediator.attach();
    }

    @Override
    public void onBackPress(View view) {
        onBackPressed();
    }

    @Override
    public void onClick(View v) {
        startActivity(new Intent(ChatroomListActivity.this, ChatroomProfileActivity.class));
    }

}

File: business/chatroom/src/main/java/com/easemob/chatroom/ui/activity/ChatroomListActivity.kt
package com.easemob.chatroom.ui.activity

import android.content.Intent
import android.graphics.Typeface
import android.os.Bundle
import android.util.Log
import android.util.TypedValue
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.alibaba.android.arouter.facade.annotation.Route
import com.easemob.baseui.BaseUiActivity
import com.easemob.buddy.tool.FastClickTools.isFastClick
import com.easemob.chatroom.R
import com.easemob.chatroom.databinding.ChatroomListLayoutBinding
import com.easemob.chatroom.general.repositories.ProfileManager
import com.easemob.chatroom.model.PageViewModel
import com.easemob.chatroom.ui.fragment.ChatroomListFragment
import com.easemob.config.RouterPath
import com.easemob.secnceui.widget.titlebar.ChatroomTitleBar
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.OnTabSelectedListener
import com.google.android.material.tabs.TabLayoutMediator
import com.hyphenate.util.DensityUtil

@Route(path = RouterPath.ChatroomListPath)
class ChatroomListActivity : BaseUiActivity<ChatroomListLayoutBinding>(),
    ChatroomTitleBar.OnBackPressListener, View.OnClickListener {
    private var pageViewModel: PageViewModel? = null
    private var title: TextView? = null
    private var mCount = 0
    private var index = 0
    private var resId = 0
    private val titles = intArrayOf(R.string.app_easemob_chatroom)

    override fun getViewBinding(inflater: LayoutInflater): ChatroomListLayoutBinding? {
        return ChatroomListLayoutBinding.inflate(inflater)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initView(savedInstanceState)
        initData()
        initListener()
    }

    private fun initListener() {
        pageViewModel = ViewModelProvider(this)[PageViewModel::class.java]
        binding.titleBar.setOnBackPressListener(this)
        binding.imageAvatar.setOnClickListener(this)

        binding.bottomLayout.setOnClickListener {
            startActivity(
                Intent(
                    this@ChatroomListActivity,
                    ChatroomCreateActivity::class.java
                )
            )
        }

        binding.listTabLayout.addOnTabSelectedListener(object : OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                if (tab.customView != null) {
                    index = tab.position
                    title = tab.customView?.findViewById<TextView>(R.id.tab_item_title)
                    title?.setTextColor(resources.getColor(R.color.dark_grey_color_040925))
                    title?.setTypeface(null, Typeface.BOLD)
                }
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {
                if (tab.customView != null) {
                    val title = tab.customView!!.findViewById<TextView>(R.id.tab_item_title)
                    title.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12f)
                    title.setText(titles[tab.position])
                    title.setTextColor(resources.getColor(R.color.color_979CBB))
                    setTextStyle(title, Typeface.NORMAL)
                }
            }

            override fun onTabReselected(tab: TabLayout.Tab) {
                title?.text = getString(titles[index])
            }
        })

        binding.vpFragment.registerOnPageChangeCallback(object : OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                if (position == 0) {
                    pageViewModel!!.setPageSelect(-1)
                } else if (position == 1) {
                    pageViewModel!!.setPageSelect(0)
                } else if (position == 2) {
                    pageViewModel!!.setPageSelect(1)
                }
            }
        })
    }

    private fun initData() {
        setupWithViewPager()
    }

    private fun initView(savedInstanceState: Bundle?) {
        setTextStyle(binding.titleBar.title, Typeface.BOLD)
        ViewCompat.setOnApplyWindowInsetsListener(binding.root) { _: View?, insets: WindowInsetsCompat ->
            val inset = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            binding.clRoot.setPaddingRelative(0, inset.top, 0, inset.bottom)
            WindowInsetsCompat.CONSUMED
        }
    }

    override fun onBackPress(view: View?) {

    }

    override fun onClick(v: View) {
        if (!isFastClick(v, 1000)) startActivity(
            Intent(
                this@ChatroomListActivity,
                ChatroomProfileActivity::class.java
            )
        )

    }

    private fun setupWithViewPager() {
        binding.vpFragment.offscreenPageLimit = ViewPager2.OFFSCREEN_PAGE_LIMIT_DEFAULT
        // set adapter
        binding.vpFragment.adapter =
            object : FragmentStateAdapter(supportFragmentManager, lifecycle) {
                override fun createFragment(position: Int): Fragment {
                    val fragment = ChatroomListFragment()
                    val bundle = Bundle()
                    bundle.putInt("position", position)
                    fragment.arguments = bundle
                    fragment.SetItemCountChangeListener { count ->
                        mCount = count
                        val layoutParams: ViewGroup.LayoutParams? = title?.layoutParams
                        layoutParams?.height = DensityUtil.dip2px(this@ChatroomListActivity, 26f)
                        title?.setTextSize(TypedValue.COMPLEX_UNIT_SP, 16f)
                        setTextStyle(title, Typeface.BOLD)
                        title?.gravity = Gravity.CENTER
                        val content = getString(titles[index]) + getString(
                            R.string.chatroom_tab_layout_count,
                            mCount.toString()
                        )
                        title?.setTextColor(getColor(R.color.dark_grey_color_040925))
                        title?.text = content
                    }
                    return fragment
                }

                override fun getItemCount(): Int {
                    return titles.size
                }
            }

        // set TabLayoutMediator
        val mediator = TabLayoutMediator(
            binding.listTabLayout, binding.vpFragment
        ) { tab, position ->
            tab.setCustomView(R.layout.tab_item_layout)
            val title = tab.customView!!.findViewById<TextView>(R.id.tab_item_title)
            title.setText(titles[position])
        }
        // setup with viewpager2
        mediator.attach()
    }

    override fun onResume() {
        super.onResume()
        try {
            resId = resources.getIdentifier(
                ProfileManager.getInstance().profile.portrait, "drawable",
                packageName
            )
            binding.imageAvatar.setImageResource(resId)
        } catch (e: Exception) {
            Log.e("getResources()", e.message.toString())
        }
    }
}